.github/workflows/CI.yml

```diff
@@ -0,0 +1,85 @@
+name: Main CI
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+    types: [opened, synchronize, reopened]
+jobs:
+  branch-naming-rules:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: deepakputhraya/action-branch-name@master
+        with:
+          regex: '^(feature|bugfix|improvement|library|prerelease|release|hotfix|poc)\/[a-z0-9_.-]+$'
+          allowed_prefixes: 'feature,bugfix,improvement,library,prerelease,release,hotfix,poc'
+          ignore: main,dev
+          min_length: 5
+          max_length: 50
+
+  golangci:
+    name: Golang CI
+    needs: [branch-naming-rules]
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/setup-go@v3
+        with:
+          go-version: '^1.19'
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+      - name: golangci-lint
+        uses: golangci/golangci-lint-action@v3
+        with:
+          version: latest
+          skip-cache: true
+          skip-pkg-cache: true
+          skip-build-cache: true
+
+  unittest:
+    name: Run unit test
+    needs: [branch-naming-rules]
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/setup-go@v3
+        with:
+          go-version: '^1.19'
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+      - name: Run test
+        run:  go test $(go list ./... | grep -v /vendor/) -coverprofile=coverage.out -v
+      - uses: actions/upload-artifact@v3
+        with:
+          name: coverage
+          path: coverage.out
+
+  testapp:
+    name: Run test app
+    needs: [unittest]
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/setup-go@v3
+        with:
+          go-version: '^1.19'
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+      - name: Run test app 
+        run: |
+          cd testapp/
+          TESTAPP_ROOTNODE=12  go run main.go --config_path .
+
+  sonarcloud:
```

Comment:
I think you have to add
`needs: [unit-tests]` and to retrieve the coverage report from unit test job